add draft

juanvallejo opened the pull request:
Rough Draft. Collecting feedback before opening publicly.

cc @deads2k @soltysh

juanvallejo committed:
add kubectl plugins KEP

juanvallejo commented:
~~This is a work in progress. Most of the sections under "Proposal" still need to be finished~~

soltysh commented on keps/sig-cli/0015-updated-plugin-mechanism.md:
```diff
@@ -0,0 +1,125 @@
+---
+kep-number: 15
+title: Updated Plugin Mechanism
```
Kubectl Plugins

juanvallejo replied:
Naming things is hard :)

soltysh commented on keps/sig-cli/0015-updated-plugin-mechanism.md:
```diff
@@ -0,0 +1,125 @@
+
+## Summary
+
+This proposal introduces the main design for an updated plugin mechanism, to replace the existing plugin system in `kubectl`.
```
What we had before was alpha, so it's as it didn't exist ;) Keep it simple by calling it just plugin mechanism.

soltysh commented on keps/sig-cli/0015-updated-plugin-mechanism.md:
```diff
@@ -0,0 +1,125 @@
+
+## Motivation
+
+The existing plugin system in `kubectl` presents a few limitations with its current design. It forces plugin scripts and executables 
```
Here would be the only place where we could state that the existing alpha mechanism proved not to be good enough. Maybe it's worth pointing to extensions people asked for in the mechanism, such as https://github.com/kubernetes/kubernetes/issues/53640

soltysh commented on keps/sig-cli/0015-updated-plugin-mechanism.md:
```diff
@@ -0,0 +1,125 @@
+### Goals
+
+- Support a simple plugin installation process (no additional config, users drop a binary in their `PATH`, for example, and they are then able to use them).
+  - No additional configuration is needed, only the plugin binary (or script, or set of scripts).
```
Maybe call it executable, since that includes both binaries as well as scripts.

soltysh commented on keps/sig-cli/0015-updated-plugin-mechanism.md:
```diff
@@ -0,0 +1,125 @@
+
+- Support a simple plugin installation process (no additional config, users drop a binary in their `PATH`, for example, and they are then able to use them).
+  - No additional configuration is needed, only the plugin binary (or script, or set of scripts).
+  - A plugin's filename determines the plugin's intention, such as which path in the command tree it applies to:
```
If possible I'd try to stay away from that bit of the proposal. I'm not 100% how we should solve it, yet.

juanvallejo replied:
I'd be okay with updating this part of the KEP as the current prototype for the mechanism evolves with feedback. Seeing that we'd like to stay away from additional config files and such, I think the [current logic that we have in the prototype](https://github.com/juanvallejo/kubernetes/pull/9#issuecomment-393627865) covers most valid use-cases with this design

soltysh commented on keps/sig-cli/0015-updated-plugin-mechanism.md:
```diff
@@ -0,0 +1,125 @@
+  - No additional configuration is needed, only the plugin binary (or script, or set of scripts).
+  - A plugin's filename determines the plugin's intention, such as which path in the command tree it applies to:
+    - `/usr/bin/kubectl-create-foobar` would, for example, override (or create) the command `kubectl create foobar`.
+- Provide a way for plugins to override specific subcommands, while preventing other paths from being overriden.
```
Same here - it's open for discussion. Let's focus on the mechanism for now and not allowing overrides.

soltysh replied:
Overrides should be a future item.

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+  - "@deads2k"
+  - "@liggitt"
+approvers:
+  - TBD
```
coordinate with Phil.  You should have a non-redhat approver

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+  - "@liggitt"
+approvers:
+  - TBD
+  - "@deads2k"
```
soltysh I would think.  We need to have more distributed power and he's a sig lead in the right sig

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+title: Kubectl Plugins 
+authors:
+  - "@juanvallejo"
+  - "@soltysh"
```
@soltysh you sure you want credit here instead of being an approver?

liggitt replied:
agree, author+approver is weird

soltysh replied:
approver, I'll be hidden author ;)

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+
+The existing alpha plugin system in `kubectl` presents a few limitations with its current design. 
+It forces plugin scripts and executables to exist in a pre-determined location, requires basic information for a single 
+plugin to be declared explicitly in its own file, anddoes not provide a clear way to override existing command paths, 
```
typo

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+## Motivation
+
+The existing alpha plugin system in `kubectl` presents a few limitations with its current design. 
+It forces plugin scripts and executables to exist in a pre-determined location, requires basic information for a single 
```
requires a per-plugin metadata file for interpretation

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+  Plugins receive all arguments and flags provided by users and are responsible for adjusting their behavior
+  accordingly.
+* Provide a way for plugins to override specific subcommands, while preventing other paths from being overriden.
+* Provide a standard package of common cli utilities that is consumed by `kubectl` and plugins alike.
```
I don't think that this goal is related to the plugin mechanism itself.

juanvallejo replied:
Thanks, will remove this line

soltysh replied:
It's reasonable to move that to non-goal. It's a nice-to-have thing that we will be working on.

deads2k commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,152 @@
+  They use a programming language of their choice and create an executable file.
+  The executable's filename consists of the command path to replace, and is prefixed with `kubectl-`.
+  The executable file is placed on the user's PATH, preferably under `/usr/local/bin`.
+  If a plugin is written using a dynamic language, and depends on logic defined in additional files, those
```
Why put this here?  Seems like extraneous detail.

juanvallejo replied:
I can move this to the "Implementation Details" section if it makes more sense there. I wanted to clarify how certain aspects of making a complex plugin, for example, would be impacted by this design

liggitt commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+`kubectl foo --baz bar`, but instead end up matching `kubectl-foo` with the flag `--baz` and the argument `bar` based
+on the placement of the flag `--baz`.
+
+A notable constraint of this design is that it excludes any form of plugin lifesycle management, or version compatibility.
```
lifecycle

liggitt commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+on the placement of the flag `--baz`.
+
+A notable constraint of this design is that it excludes any form of plugin lifesycle management, or version compatibility.
+A plugin may depend on other plugins based on the desicion of a plugin author, however the current design does nothing
```
decision

liggitt commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+For example, `kubectl get foobar` could now be failing due to:
+- An actual bug with the `get` command
+- `foobar` might be a CRD that was not defined correctly
+- There could be a plugin overriding the path `kubectl-get-foobar` and that plugin is experiencing an error.
```
is this design proposing allowing plugins to override every kubectl command/subcommand? is there a reason we wouldn't start with the existing plugin extension point and maybe the create command? letting a plugin override `kubectl get foobar` seems like a non-goal to me

juanvallejo replied:
> is this design proposing allowing plugins to override every kubectl command/subcommand? 

No, at least it is not my intention for this to propose overriding any path. I meant to use this point as a means to justify the last paragraph in this section, which calls for a way to only allow plugins to override an allowed set of command paths.

soltysh replied:
During one of the earlier sig-cli meetings we've agreed to start with not allowing to override the defaults. That will come as a natural evolution.

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+This proposal introduces the main design for a plugin mechanism in `kubectl`.
+The mechanism is a git-style system, that looks for executables on a user's `$PATH` whose name begins with `kubectl-`.
+This design allows plugin binaries to override specific command paths, replacing allowed subcommands in existing commands,
+or adding new subcommands under the `kubectl plugin` path.
```
I thought we wanted not to have the `plugin` subpath, but rather directly under kubectl. So that `kubectl-foo` becomes `kubectl foo`

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+
+This proposal introduces the main design for a plugin mechanism in `kubectl`.
+The mechanism is a git-style system, that looks for executables on a user's `$PATH` whose name begins with `kubectl-`.
+This design allows plugin binaries to override specific command paths, replacing allowed subcommands in existing commands,
```
This ~~design~~ allows

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+or provide additional subcommands without having to override a top-level command.
+
+A re-design plugins system allows us to implement extensibility requests from users that the current system cannot address
+(See https://github.com/kubernetes/kubernetes/issues/53640).
```
There's one more https://github.com/kubernetes/kubernetes/issues/55708

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+
+### Goals
+
+* Support a simple plugin installation process (no additional config, users drop a binary in their `PATH`, for example, 
```
We don't want to support any installation process, we just pick up what's available in the `$PATH`.

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+* Developer wants to create and expose a plugin to `kubectl`.
+  They use a programming language of their choice and create an executable file.
+  The executable's filename consists of the command path to replace, and is prefixed with `kubectl-`.
+  The executable file is placed on the user's PATH, preferably under `/usr/local/bin`.
```
Drop the preferably... we don't care.

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+  The executable file is placed on the user's PATH, preferably under `/usr/local/bin`.
+  If a plugin is written using a dynamic language, and depends on logic defined in additional files, those
+  files are placed elsewhere in the user's system (e.g. ~/.kubeplugins/myplugin/helper1.py) and not in the user's `PATH`.
+* User wants to "install" a plugin `kubectl-plugin-foo`
```
Installation is out of scope. This is only for running them.

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+  files are placed elsewhere in the user's system (e.g. ~/.kubeplugins/myplugin/helper1.py) and not in the user's `PATH`.
+* User wants to "install" a plugin `kubectl-plugin-foo`
+  They move the executable file `kubectl-plugin-foo` to a location on their path, and access the plugin via the command `kubectl plugin foo`.
+* User wants a subset of their plugin executables to live in a specific location on their filesystem.
```
Not relevant to plugin mechanism itself.

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+* User wants to "install" a plugin `kubectl-plugin-foo`
+  They move the executable file `kubectl-plugin-foo` to a location on their path, and access the plugin via the command `kubectl plugin foo`.
+* User wants a subset of their plugin executables to live in a specific location on their filesystem.
+  The user creates a directory in the desired location (`~/Documents/foo-plugins`) and adds that location to their `PATH`.
```
Not relevant.

soltysh commented on keps/sig-cli/0015-kubectl-plugins.md:
```diff
@@ -0,0 +1,151 @@
+and the arguments `--bar baz` will be passed to it in that order.
+
+A potential limitation of this could present itself in the order of arguments provided by a user.
+A user could intend to run a plugin `kubectl-plugin-foo-bar` with the flag `--baz` with the following command
```
Arguments are not under consideration when matching plugin name,  so this is not a limitation here.

juanvallejo replied:
I would say the limitation would still exist in the constraint that is placed in the order in which a user can provide subcommands and flags. Although the plugin mechanism does not handle or parse any args meant for a plugin, the plugin itself might not be matched if a --flag is specified before its full path

juanvallejo commented:
- Updated the "Risks and Mitigations" section to no longer mention overriding existing command paths, but rather to mention the "land rush" effect that might take place since we are no longer requiring commands provided by plugins to exist within the `kubectl plugin` space
- Updated the "Summary" section to no longer mention the `kubectl plugin` space
- Addressed further comments

soltysh reviewed (approved):
Add some pseudo-code we've talked and this is good.

juanvallejo commented:
Closing. This has been opened against the upstream community repo.
